### DISCUSSION TITLE
ImmutableCommand in memstate?

### DISCUSSION THREAD

**Discussion by InspiringCode** [DevrexLabs/memstate#122] [Q&A]
Hi,

I have another question. Do you support immutable command and immutable models in memstate? The beta documentation mentions them, but I guess it just applies to OrigoDB.

If not, how do you ensure consistent results when commands and queries are executed concurrently?

**Comment by rofr** [comment]
An immutable model means that the commands return the next version of the model. Meanwhile, queries are free to complete on the version that was current when they started executing. OrigoDB has this as an option. I considered going all in with immutable modeling in memstate but wasn't brave enough to make that decision.

For normal mode (the only mode in Memstate), the engine uses a ReaderWriterLockSlim, which allows EITHER multiple simultaneous readers or a single writer.

**Comment by InspiringCode** [comment]
Aw, what a pity that you didn't go for immutable models, would make perfect sense in my opinion cause it saves one from so many troubles.

Do you plan to add back immutability support? Would it even be possible with the current architecture?

**Reply by rofr** [reply]
Performance of the immutable collections libraries I looked at is pretty poor. And if the model is complex/deep, writing commands becomes painful. I think the architecture will allow it as an option later. It's on the roadmap but don't count on it happening. I'm actually considering a memstate.js built on redux as a future project.

And I agree that immutable models are really elegant and eliminate some hard problems altogether. When a command crashes (throws an exception) there is a possibility that the model has been corrupted, so A in ACID is the command programmers responsibility. With immutability, there is no new model returned so nothing is changes. This is guaranteed by the engine.

Another huge benefit is that queries don't have to wait for commands. They start executing immediately on the current version of the model, even if a command is churning away. Normally commands execute instantly but in theory the entire system can be blocked by a rogue command.

**Reply by InspiringCode** [reply]
Hmm. I think that the System.Collections.Immutable are highly optimized and perform quite well if you choose the appropriate collection type but I have to admit that I don't have much experience with them.

I think that with the new record types in C# 9 and especially with the "[with expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression)".

**Reply by InspiringCode** [reply]
But yes, I would also love to see more work, research and projects be done on the area of immutability in C#. If you look for immutable collections for example, you don't really have a choice...
